File: api/models/note.go
package models

type Note struct {
	Id        uint64 `json:"id" db:"id"`
	SkillId   uint64 `json:"skillId" db:"skillId"`
	ProjectId uint64 `json:"projectId" db:"projectId"`
	UserID    uint64 `json:"userId" db:"userId"`
	Content   string `json:"content" db:"content"`
}

func NewNote(
	id uint64,
	skillId uint64,
	projectId uint64,
	userID uint64,
	content string,
) Note {
	return Note{
		Id:        id,
		SkillId:   skillId,
		ProjectId: projectId,
		UserID:    userID,
		Content:   content,
	}
}

type NoteDetail struct {
	Id      uint64 `json:"id" db:"id"`
	Skill   string `json:"skill"`
	Project string `json:"project"`
	User    string `json:"user"`
	Content string `json:"content" db:"content"`
}

func NewNoteDetail(
	id uint64,
	skill string,
	project string,
	user string,
	content string,
) NoteDetail {
	return NoteDetail{
		Id:      id,
		Skill:   skill,
		Project: project,
		User:    user,
		Content: content,
	}
}

type CreateNote struct {
	SkillId   uint64 `json:"skillId" db:"skillId"`
	ProjectId uint64 `json:"projectId" db:"projectId"`
	UserID    uint64 `json:"userId" db:"userId"`
	Content   string `json:"content" db:"content"`
}

func NewCreateNote(
	skillId uint64,
	projectId uint64,
	userID uint64,
	content string,
) CreateNote {
	return CreateNote{
		SkillId:   skillId,
		ProjectId: projectId,
		UserID:    userID,
		Content:   content,
	}
}

type NoteFilter struct {
	Skill   *string `json:"skill"`
	Project *string `json:"project"`
	User    *string `json:"user"`
}

func NewNoteFilter(
	skill *string,
	project *string,
	user *string,
) NoteFilter {
	return NoteFilter{
		Skill:   skill,
		Project: project,
		User:    user,
	}
}

File: web/src/services/note.ts
import { NoteSchema } from "../schemas/note";

const API_URL = import.meta.env.VITE_API_URL;

/* eslint-disable  @typescript-eslint/no-explicit-any */
export function mapNotes(data: any): NoteSchema[] {
  return data
    .map((note: any) => mapNote(note))
    .filter((note: NoteSchema | undefined) => !!note);
}

/* eslint-disable  @typescript-eslint/no-explicit-any */
export function mapNote(data: any): NoteSchema | undefined {
  if (!data) return undefined;
  return {
    id: data.id,
    skill: data.skill,
    projectId: data.projectId,
    userId: data.userId,
    content: data.content,
  };
}

export async function getNotesOfUserProject(
  userId: number,
  projectId: number,
): Promise<NoteSchema[]> {
  if (userId === 0 || projectId === 0) {
    return [];
  }
  return fetch(`${API_URL}/notes?user=${userId}&project=${projectId}`)
    .then((response) => response.json())
    .then((data) => mapNotes(data));
}
